diff --git a/core/comboot.inc b/core/comboot.inc
index e5dc115..60506ce 100644
--- a/core/comboot.inc
+++ b/core/comboot.inc
@@ -1087,6 +1087,7 @@ comboot_hook_entry:
 		push fs
 		push es
 		push ds
+		mov ecx,[KernelSize]
 		call far [comboot_far]
 		pop ds
 		pop es
diff --git a/core/ui.inc b/core/ui.inc
index 59b5724..d56c148 100644
--- a/core/ui.inc
+++ b/core/ui.inc
@@ -612,7 +612,11 @@ kernel_good:
 		mov [KernelCNameLen],di
 
 		; Default memory limit, can be overridden by image loaders
+		mov eax,[AltHighMemSize]
+		or eax,eax
+		jnz .altsize
 		mov eax,[HighMemRsvd]
+.altsize:
 		mov [MyHighMemSize],eax
 
 		popad
@@ -633,6 +637,7 @@ kernel_good:
 ; At this point, EAX contains the size of the kernel, SI contains
 ; the file handle/cluster pointer, and ECX contains the extension (if any.)
 ;
+		mov [KernelSize],eax
 		movzx di,byte [KernelType]
 		add di,di
 		jmp [kerneltype_table+di]
