diff --git a/core/runkernel.inc b/core/runkernel.inc
index 5da5bc9..033019c 100644
--- a/core/runkernel.inc
+++ b/core/runkernel.inc
@@ -168,7 +168,7 @@ opt_mem:
 		ret
 
 opt_quiet:
-		mov byte [QuietBoot],QUIET_FLAG
+		or byte [QuietBoot],QUIET_FLAG
 		ret
 
 %if IS_PXELINUX
@@ -228,7 +228,9 @@ new_kernel:
 ; we were provided.
 ;
                 mov al,[es:su_loadflags]
-		or al,[QuietBoot]		; Set QUIET_FLAG if needed
+		mov ah,[QuietBoot]		; Set QUIET_FLAG if needed
+		and ah,QUIET_FLAG
+		or al,ah
 		mov [es:su_loadflags],al
 		mov [LoadFlags],al
 
@@ -642,7 +644,7 @@ loadinitrd:
 ;		assumes CS == DS
 ;
 writestr_qchk:
-		test byte [QuietBoot],QUIET_FLAG
+		test byte [QuietBoot],QUIET_FLAG | 2
 		jz writestr
 		ret
 
@@ -697,4 +699,6 @@ KernelVersion	resw 1			; Kernel protocol version
 ;
 InitRDPtr	resw 1			; Pointer to initrd= option in command line
 LoadFlags	resb 1			; Loadflags from kernel
-QuietBoot	resb 1			; Set if a quiet boot is requested
+
+		section .data
+QuietBoot	db 0			; Set if a quiet boot is requested
diff --git a/core/ui.inc b/core/ui.inc
index 0d896fc..a256ed0 100644
--- a/core/ui.inc
+++ b/core/ui.inc
@@ -616,7 +616,7 @@ kernel_good:
 		;
 		xor ax,ax
 		mov [InitRDPtr],ax
-		mov [QuietBoot],al
+		and byte [QuietBoot],~QUIET_FLAG
 %if IS_PXELINUX
 		mov [KeepPXE],al
 %endif
